### PR TITLE
Support Doctrine DBAL 4

### PR DESCRIPTION
**Pull request by stloyd**
<h2>Change Log</h2>
<div id="change-log">
  <h4>Added</h4>
  <ul id="added">
    <li>Support Doctrine DBAL 4</li>
  </ul> 
  <h4>Fixed</h4>  
  <ul id="fixed">
    
  </ul>
  <h4>Changed</h4>
  <ul id="changed">
    
  </ul>  
  <h4>Removed</h4>
  <ul id="removed">
    
  </ul>
  <h4>Deprecated</h4>
  <ul id="deprecated">
    
  </ul>  
  <h4>Security</h4> 
  <ul id="security">
    
  </ul>     
</div>
<hr/>

<h2>Description</h2>

### CODE REVIEW

File: src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/Parameter.php [hunk 1]
```diff
@@ -13,7 +13,7 @@ final class Parameter implements QueryParameter
     public function __construct(
         public readonly string $queryParamName,
         public readonly EntryReference $ref,
-        public readonly int $type = ArrayParameterType::STRING,
+        public readonly ArrayParameterType $type = ArrayParameterType::STRING,
```
Comment (stloyd) [inline src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/Parameter.php]: In 4 it's an enum class, not an integer.

File: src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Integration/MySqlBulkInsertTest.php [hunk 2]
```diff
@@ -23,7 +23,7 @@ public function test_inserts_multiple_rows_at_once() : void
                     new Column('id', Type::getType(Types::STRING), ['notnull' => true, 'length' => 255]),
                     new Column('age', Type::getType(Types::INTEGER), ['notnull' => true]),
                     new Column('name', Type::getType(Types::STRING), ['notnull' => true, 'length' => 255]),
-                    new Column('description', Type::getType(Types::STRING), ['notnull' => false]),
+                    new Column('description', Type::getType(Types::STRING), ['notnull' => false, 'length' => 255]),
```
Comment (stloyd) [inline src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Integration/MySqlBulkInsertTest.php:26]: In v4 setting `length` is mandatory.

**Comment by norberttech** [comment]
is 3.x no longer supported by doctrine?

File: src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Integration/MySqlBulkInsertTest.php [hunk 3]
```diff
@@ -36,9 +36,9 @@ public function test_inserts_multiple_rows_at_once() : void
             $this->databaseContext->connection(),
             $table,
             new BulkData([
-                ['id' => $id1 = \bin2hex(\random_bytes(5)), 'age' => 20, 'name' => 'Name One', 'description' => 'Description One', 'active' => false, 'updated_at' => $date1 = new \DateTimeImmutable(), 'tags' => \json_encode(['a', 'b', 'c'])],
-                ['id' => $id2 = \bin2hex(\random_bytes(5)), 'age' => 30, 'name' => 'Name Two', 'description' => null, 'active' => true, 'updated_at' => $date2 = new \DateTimeImmutable(), 'tags' => \json_encode(['a', 'b', 'c'])],
-                ['id' => $id3 = \bin2hex(\random_bytes(5)), 'age' => 40, 'name' => 'Name Three', 'description' => 'Description Three', 'active' => false, 'updated_at' => $date3 = new \DateTimeImmutable(), 'tags' => \json_encode(['a', 'b', 'c'])],
+                ['id' => $id1 = \bin2hex(\random_bytes(5)), 'age' => 20, 'name' => 'Name One', 'description' => 'Description One', 'active' => false, 'updated_at' => $date1 = new \DateTime(), 'tags' => \json_encode(['a', 'b', 'c'])],
```
Comment (stloyd) [inline src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Integration/MySqlBulkInsertTest.php:39]: This looks like a bug in DBAL, but without this change, it ends with an error that `DateTimeType` (from DBAL) cannot accept the `DateTimeImmutable` instance.

File: src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Unit/BulkDataTest.php [hunk 4]
```diff
@@ -99,11 +100,11 @@ public function test_returns_all_sql_parameters_as_one_dimensional_array_with_pl
             $bulkData->toSqlParameters(
                 new TableDefinition(
                     'test',
-                    new Column('date', new StringType()),
-                    new Column('title', new StringType()),
-                    new Column('description', new StringType()),
-                    new Column('quantity', new IntegerType()),
-                    new Column('errors', new JsonType()),
+                    new Column('date', $registry->get(Types::STRING)),
```
Comment (stloyd) [inline src/lib/doctrine-dbal-bulk/tests/Flow/Doctrine/Bulk/Tests/Unit/BulkDataTest.php:103]: This is required due to the fact new registry holds the hash of type objects, and creating it here returns a different hash which ends with an exception due to mismatch.

**Comment by stloyd** [comment]
@norberttech Version 3.8+ is right now supported, but in v4 there are plenty of hard BC breaks, that probably would need to end up in duplicating code to support both versions.

**Comment by norberttech** [comment]
> @norberttech Version 3.8+ is right now supported, but in v4 there are plenty of hard BC breaks, that probably would need to end up in duplicating code to support both versions.

Yeah, then we need to figure out way to keep 3.x supported

File: src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/DbalQueryExtractor.php [hunk 5]
```diff
@@ -42,6 +40,11 @@ public static function single(Connection $connection, string $query, array $para
     public function extract(FlowContext $context) : \Generator
     {
         foreach ($this->parametersSet->all() as $parameters) {
+            /**
+             * @phpstan-ignore-next-line
+             *
+             * @psalm-suppress InvalidArgument
+             */
```
Comment (stloyd) [inline src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/DbalQueryExtractor.php:47]: This is required due to phpdoc change between 3.x & 4.x, in legacy there is `int` (legacy approach in `ArrayParameterType`).

File: src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/Parameter.php [hunk 6]
```diff
@@ -13,7 +13,7 @@ final class Parameter implements QueryParameter
     public function __construct(
         public readonly string $queryParamName,
         public readonly EntryReference $ref,
-        public readonly int $type = ArrayParameterType::STRING,
+        public readonly int|ArrayParameterType $type = ArrayParameterType::STRING,
```
Comment (stloyd) [inline src/adapter/etl-adapter-doctrine/src/Flow/ETL/Adapter/Doctrine/Parameter.php]: @norberttech Question here, as we have getters, shouldn't those be `private`?

File: src/lib/doctrine-dbal-bulk/composer.json [hunk 7]
```diff
@@ -12,7 +12,7 @@
     ],
     "require": {
         "php": "~8.1.0 || ~8.2.0 || ~8.3.0",
-        "doctrine/dbal": "^3.6"
+        "doctrine/dbal": "^3.8 || ^4.0"
```
Comment (norberttech) [inline src/lib/doctrine-dbal-bulk/composer.json]: how much of an effort it would be to keep ^3.6?

Comment (stloyd) [reply src/lib/doctrine-dbal-bulk/composer.json]: Could be not too big, but that version is not maintained anymore.

**Comment by norberttech** [comment]
nice 🚀

**Pull request merged**